README.md
# SenseMe
Python Library for Haiku SenseMe app controlled fans/lights

This library is useful to control Haiku SenseMe fans with light kits. I would expect it will probably also work with their lights.

It might also be useful for controlling DIY projects as the protocol is very simple and would be easy to clone. And, if you were to use their API the Android and iOS apps may work to control DIY devices. A suggested idea would be to add an Arduino or Raspberry Pi and a relay or two to your own fan and use this or the Haiku Home app to control them. I've read that doing this sort of thing [with a resistor is a dangerous idea, so don't do that.](https://arstechnica.com/civis/viewtopic.php?t=1263401)

Sniffing the packets and documenting the protocol were the work of [Bruce](http://bruce.pennypacker.org/tag/senseme-plugin/). Much of the code was based on his work on making [an Indigo plugin](https://github.com/bpennypacker/SenseME-Indigo-Plugin)

See [Issues](issues) for known issues or if you want to contribute but don't know where to start. Some easy fixes are labeled.

I am not affiliated with Haiku Home or Big Ass Fans. Their support rep said this project seemed cool in it's infancy, and they even answered a technical question regarding the protocol for me, so hopefully they still approve.

# Future
Some ideas for future related projects and features:

 1. Plex plugin (dim the fans when the movie starts and light up when it is paused or ends? Yes, please!)

 2. Alexa / Google Home plugins (see REST API below)

 3. Store information in a database (sqlite, or even json, would be fine) rather than discovering each time.

 4. Track usage and temperatures

 5. Other automation system plugins

 6. REST API backend with a CLI client (a basic direct to fan command line client was added in 0.1.3)

 7. React or other web based client for the above API

 8. Docker image for the above

 9. Use protocol to control other devices by attaching an Arduino or Pi

 10. Discover the rest of the protocol. (Run strings on the apps and some packet sniffing.)

 11. More examples

 12. ~~Command Line Client~~ Basic client added in 0.1.3, see [SenseMe CLI](#cli) below
 
 13. Unit Tests


# Usage
    from senseme import discover
    # discover devices, returns list of SenseMe devices
    devices = discover()
    fan = devices[0]
    repr(fan)

SenseMe(name='Living Room Fan', ip='192.168.1.50', model='FAN,HAIKU', series='HSERIES', mac='20:F8:5E:E3:AB:00')


    # Statically assign the fan? Probably not, but you would do it this way:
    from senseme import SenseMe
    fan = SenseMe('192.168.1.50', 'Living Room Fan', model='FAN')
    # or, this might be easier
    fan = SenseMe(name="Living Room Fan")

Control the fan:

    # Turn the light off / on
    fan.light_powered_on = False
    fan.light_powered_on = True
    # or, if you just want to toggle it
    light_status = fan.light_toggle()

    # Increase light level by 2 levels
    fan.inc_brightness(2)

    # Get Light level
    print(fan.brightness)

    # Fan Speeds
    fan.fan_powered_on = True
    fan.fan_powered_on = False
    print(fan.speed)

    # whoosh mode
    fan.whoosh = True

    # want an increasing light effect? Do this.
    # But, really, probably don't, I don't think Haiku intended strobe effects.
    # I'm not responsible if you make a strobe light and break the fan or worse
    for intensity in range(1,16):
        fan.brightness = intensity
        time.sleep(1)

    # export details to json / xml / str(dict)
    fan.json
    fan.xml
    fan.dict  # nested dict
    fan.flat_dict  # flattened

    # Listen for broadcasts, useful for debugging,
    # wouldn't suggest using it for anything else
    fan.listen()

# SenseMe CLI <a id="cli"></a>
In version 0.1.3 a script was added to control HaikuHome SenseMe devices from a command line.

Set the following environment variables:

    SENSEME_DEFAULT_FAN_IP
    SENSEME_DEFAULT_FAN_NAME

Then run `senseme_cli` with one of the following options.

    fan x           - set speed to x (0-7)
    fan on|off      - fan on/off
    light x         - set light brightness to x
    light on|off    - light on/off
    whoosh [on|off] - whoosh mode on/off

# Requirements
Python 3.6 is required to use this library.
SenseMe devices communicate on UDP port 31415 using broadcast and unicast traffic. Any firewall will have to allow this through.



CHANGES.txt
v0.1.0, 2018-03-06 -- Initial release.
v0.1.1, 2018-03-12 -- Fix neglected lib directory
v0.1.2, 2018-06-23 -- Fix issues with reading fan variables
v0.1.3, 2018-07-05 -- Add command line script (senseme_cli), improve setup.py
v0.1.4, 2018-07-09 -- Prevent setup.py install installation on 2.7


sensemefan.py
import socket
import re

import time

'''
class SenseMeFan:
This class provides UDP access to Haiku SenseMe capable fans.
Based on work from Bruce at http://bruce.pennypacker.org/tag/senseme-plugin/
https://github.com/bpennypacker/SenseME-Indigo-Plugin
'''

class SenseMeFan:

    def __init__(self, ip='', name='', model='', series=''):
        self.PORT = 31415

        if not ip or not name:
            self.discover()
        else:
            self.ip = ip
            self.name = name
            self.mac = ''
            self.details = ''
            self.model = model
            self.series = series

        self.light = {'brightness': None, 'status': None}
        self.fan = {'speed': None, 'status': None}

        #self.getstate()

    def __sendcommand__(self, msg):
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.sendto(msg.encode('utf-8'), (self.ip, self.PORT))
        return

    def __query__(self, msg):
        sock = socket.socket()
        sock.settimeout(5)

        sock.connect((self.ip, 31415))
        sock.send(msg.encode('utf-8'))

        try:
            status = sock.recv(1048).decode('utf-8')
	    print('Status: ' + status)
        except socket.timeout:
            print('Socket Timed Out')
        else:
            sock.close()
            print(str(status))
            matchObj = re.match('\(.*;([^;]+)\)', status)
            if matchObj:
                return matchObj.group(1)
            else:
                return False

    def setspeed(self, speed):
        if speed > 7: # max speed is 7, fan corrects to 7
            speed = 7
        elif speed < 0: # 0 also sets fan to off automatically
            speed = 0
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.__sendcommand__('<%s;FAN;SPD;SET;%s>' % (self.name, speed))
        	return
	else:
		print('Device Not Supported Yet')

    def incspeed(self, incspeed = 1):
        self.getfan()
        self.setspeed(int(self.fan['speed']) + incspeed)
        return


    def decspeed(self, decspeed = 1):
        self.getfan()
        self.setspeed(int(self.fan['speed']) - decspeed)
        return

    def setlight(self, light):
        if light > 16: # max light level, if receiving > 16 fan auto changes to 16
            light = 16
        elif light < 0: # light 0 also automatically sets pwr = off
            light = 0
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.__sendcommand__('<%s;LIGHT;LEVEL;SET;%s>' % (self.name, light))
        	return
	else:
		print('Device Not Supported Yet')

    def inclight(self, incbright = 1):
        self.getlight()
        self.setlight(int(self.light['brightness']) + incbright)
        return

    def declight(self, decbright = 1):
        self.getlight()
        self.setlight(int(self.light['brightness']) - decbright)
        return

    def fanoff(self):
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):	
        	self.__sendcommand__('<%s;FAN;PWR;OFF>' % self.name)
        	return
	else:
		print('Device Not Supported Yet')

    def fanon(self):
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
       		self.__sendcommand__('<%s;FAN;PWR;ON>' % self.name)
        	return
	else:
		print('Device Not Supported Yet')

    def fantoggle(self):
        self.getfan()
        if self.fan['status'] == 'ON':
            self.fanoff()
            return 'OFF'
        else:
            self.fanon()
            return 'ON'

    def lightoff(self):
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.__sendcommand__('<%s;LIGHT;PWR;OFF>' % self.name)
        	return
	else:
		print('Device Not Supported Yet')

    def lighton(self):
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.__sendcommand__('<%s;LIGHT;PWR;ON>' % self.name)
        	return
	else:
		print('Device Not Supported Yet')

    def lighttoggle(self):
        self.getlight()
        if self.light['status'] == 'ON':
            self.lightoff()
            return 'OFF'
        else:
            self.lighton()
            return 'ON'

    def getlight(self):
	# Commands for LSERIES fan only 
	if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.light['brightness'] = self.__query__('<%s;LIGHT;LEVEL;GET;ACTUAL>' % self.name)
        	self.light['status'] = self.__query__('<%s;LIGHT;PWR;GET>' % self.name)
        	return self.light
	else:
		print('Device Not Supported Yet')

    def getfan(self):
	# Commands for LSERIES fan only
        if ( self.model == 'FAN' and self.series == 'LSERIES' ):
        	self.fan['speed'] = self.__query__('<%s;FAN;SPD;GET;ACTUAL>' % self.name)
        	self.fan['status'] = self.__query__('<%s;FAN;PWR;GET>' % self.name)
        	return self.fan
	else:
		print('Device Not Supported Yet')

    def getstate(self):
        self.getfan()
        self.getlight()

    @property
    def fan_speed(self):
        return self.getfan()

    def listen(self):
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.bind(('', 31415))
        for x in range(1, 30):
            m = sock.recvfrom(1024)
            #print(m)

    def discover(self):
        data = '<ALL;DEVICE;ID;GET>'.encode('utf-8')
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.bind(('', 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

        res = s.sendto(data, ('<broadcast>', self.PORT))
        print(res)
        print('sent broadcast')

        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.bind(('', 31415))
        m = s.recvfrom(1048)
        print(m)
        if not m:
            pass
        else:
            self.details = m[0].decode('utf-8') 
            res = re.match('\((.*);DEVICE;ID;(.*);(.*),(.*)\)',self.details)
            self.name = res.group(1)
            self.mac = res.group(2)
            self.model = res.group(3)
            self.series = res.group(4)
            self.ip = m[1][0]
            # print('ip: ' + self.ip)
            # print('details: ' + self.details)
            # print(self.name)
        pass
